### PR TITLE
Fixing labels, should throw syntax errors when labels are followed by lexical declarations Fixes #352

### LINKED ISSUE #352: Lexical Declarations cannot appear after a label
ChakraCore currently allows lexical declarations (let, const, class, function*) to appear after a label.  Labels can only prepend a Statement, not a Declaration, according to the ES spec.

See grammar at https://tc39.github.io/ecma262/#sec-ecmascript-language-statements-and-declarations

FunctionDeclaration gets an exception and is allowed, presumably for back-compat.  GeneratorDeclaration does not get the exception, however, and so it not allow. See [LabelledItem](https://tc39.github.io/ecma262/#prod-LabelledItem)

Async functions presumably will also be disallowed by way of being a HoistableDeclaration and not excepted by LabelledItem.

E.g. each of these lines should be a syntax error but are not currently:

``` javascript
a: let x;
b: const y = 0;
c: class z { }
d: function* w() { }
e: async function u() { } // pending finalization of async function spec
```

### PR DESCRIPTION
**Pull request by kfukuda2**
Fixes #352

### CODE REVIEW

File: test/Error/Labels.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            assert.throws(() => eval("b: const y = 0;"), SyntaxError, 
+            "A const declaration must not be labelled", "Labels not allowed before lexical declaration");
+            assert.throws(() => eval("c: class z {}"), SyntaxError, 
+            "An class declaration must not be labelled", "Labels not allowed before class declaration");
+            assert.throws(() => eval("d: function* w() {}"), SyntaxError, 
+            "A generator must not be labelled", "Labels not allowed before generator declaration");
+            assert.throws(() => eval("e: async function u() { }"), SyntaxError, 
+            "An async function must not be labelled", "Labels not allowed before async function declaration");
+            /*assert.throws(() => eval("e: async function* u() { }"), SyntaxError, //TODO: Uncomment when async generators supported.
+            "An async generator must not be labelled", "Labels not allowed before async function declaration");*/
```
Comment (jackhorton) [inline test/Error/Labels.js]: There was a mention in the original bug that labels aren't allowed at the end of the program. Does this fix that? If so, could a test be added for it? `eval("var x = 0; a:")` and `eval("a:")` should throw like the rest.

File: test/Error/Labels.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        }
+    },
+    {
+        name : "Labels followed by non-declarations are valid syntax (non-generator non-async function declarations and non-lexical variable declarations are allowed)",
+        body : function () 
+        {
+            assert.doesNotThrow(() => eval(testFunctionAfterLabel), "Function declarations should be allowed after a label.");
```
Comment (tcare) [inline test/Error/Labels.js]: An interesting note about this is https://tc39.github.io/ecma262/#sec-labelled-function-declarations

Strict mode should throw, and outside of browser code non-strict should also throw. I don't think we do much distinction about whether code is running in the browser or not, apart from browser specific constructs such as event handlers. I would assume we want to allow the most relaxed rule for compat for all hosts. @digitalinfinity @boingoing and others might have an opinion here, especially in Node scenarios.

File: test/Error/Labels.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        body : function () 
+        {
+            assert.doesNotThrow(() => eval(testFunctionAfterLabel), "Function declarations should be allowed after a label."); 
+            assert.doesNotThrow(() => eval(testVarAfterLabel), "var declarations should be allowed after a label.");
+            assert.doesNotThrow(() => eval(testForLoopAfterLabel), "for loop statements should be allowed after a label.");
+            assert.doesNotThrow(() => eval(testNonLexicalDeclarationAfterLabel), "non-lexical declarations should be allowed after a label.");
+            assert.doesNotThrow(() => eval(testWhileLoopAfterLabel), "while loops should be allowed after a label");
+        }
+    }
+];
```
Comment (tcare) [inline test/Error/Labels.js]: Would you mind adding tests for some of the strict mode specific stuff? E.g.

https://tc39.github.io/ecma262/#sec-labelled-function-declarations
Note
The early error rules for WithStatement, IfStatement, and IterationStatement prevent these statements from containing a labelled FunctionDeclaration in non-strict code.

Comment (tcare) [reply test/Error/Labels.js]: Some other interesting cases are in https://tc39.github.io/ecma262/#sec-identifiers-static-semantics-early-errors, e.g. yield and async as label names depending on context and/or strict mode.

File: test/Error/rlexe.xml [hunk 4]
```diff
@@ -178,4 +178,10 @@
       <compile-flags>-ExtendedErrorStackForTestHost -force:DeferParse</compile-flags>
     </default>
   </test>
+  <test>
+    <default>
+      <files>Labels.js</files>
```
Comment (tcare) [inline test/Error/rlexe.xml]: I'm also curious if we have test coverage for these sorts of things elsewhere. If we do it probably dates back to ES5 and they may need to be updated or merged with this.

Comment (kfukuda2) [reply test/Error/rlexe.xml]: There seem to be some label tests in core/test/Basics, these could all be condensed into one test file using the unit test framework, what do you think?

Comment (tcare) [reply test/Error/rlexe.xml]: That'd be awesome, those are probably ancient and need re-evaluation anyway.

File: test/Error/Labels.js (hunk 3 above)
Comment (kfukuda2) [reply test/Error/Labels.js]: Added test cases for strict mode

File: test/Error/Labels.js (hunk 1 above)
Comment (kfukuda2) [reply test/Error/Labels.js]: Thank you, I have added the fix for that and a test

File: test/UnitTestFramework/UnitTestFramework.js [hunk 5]
```diff
@@ -112,6 +112,11 @@ var helpers = function helpers() {
         {
             return Object.prototype.toString.call(object);
         },
+        
+        getFileAndLineInfo: function getFileAndLineInfo()
```
Comment (kfukuda2) [inline test/UnitTestFramework/UnitTestFramework.js]: unittestfwerrormessages Changed the printouts from this file to give the debugger more context into the problem and be less confusing to read. 

Changes: 
- Shows the file and line of the failure. 
- asserts that differ in expectation are now shown in a way that is easier to compare output to expectation

**Pull request merged**